src/libAtomVM/nifs.c

```diff
@@ -3505,6 +3557,363 @@ static term nif_atomvm_read_priv(Context *ctx, int argc, term argv[])
+            posix_flags |= O_TRUNC;
+        // SUSv4/2018 edition flags
+#if HAVE_O_EXEC
+        } else if (globalcontext_is_term_equal_to_atom_string(glb, flag, O_EXEC_ATOM_STR)) {
+            posix_flags |= O_EXEC;
+#endif
+#if HAVE_O_SEARCH
+        } else if (globalcontext_is_term_equal_to_atom_string(glb, flag, O_SEARCH_ATOM_STR)) {
+            posix_flags |= O_SEARCH;
+#endif
+#if HAVE_O_CLOEXEC
+        } else if (globalcontext_is_term_equal_to_atom_string(glb, flag, O_CLOEXEC_ATOM_STR)) {
+            posix_flags |= O_CLOEXEC;
+#endif
+#if HAVE_O_DIRECTORY
+        } else if (globalcontext_is_term_equal_to_atom_string(glb, flag, O_DIRECTORY_ATOM_STR)) {
+            posix_flags |= O_DIRECTORY;
+#endif
+#if HAVE_O_DSYNC
+        } else if (globalcontext_is_term_equal_to_atom_string(glb, flag, O_DSYNC_ATOM_STR)) {
+            posix_flags |= O_DSYNC;
+#endif
+#if HAVE_O_NOFOLLOW
+        } else if (globalcontext_is_term_equal_to_atom_string(glb, flag, O_NOFOLLOW_ATOM_STR)) {
+            posix_flags |= O_NOFOLLOW;
+#endif
+#if HAVE_O_RSYNC
+        } else if (globalcontext_is_term_equal_to_atom_string(glb, flag, O_RSYNC_ATOM_STR)) {
+            posix_flags |= O_RSYNC;
+#endif
+#if HAVE_O_TTY_INIT
+        } else if (globalcontext_is_term_equal_to_atom_string(glb, flag, O_TTY_INIT_ATOM_STR)) {
+            posix_flags |= O_TTY_INIT;
+#endif
+        } else {
+            RAISE_ERROR(BADARG_ATOM);
+        }
+        flags = term_get_list_tail(flags);
+    }
+
+    int ok;
+    char *path = interop_term_to_string(path_term, &ok);
+    if (UNLIKELY(!ok)) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    term result;
+    int fd;
+    if (argc == 3) {
+        term mode_term = argv[2];
+        if (UNLIKELY(!term_is_integer(mode_term))) {
+            free(path);
+            RAISE_ERROR(BADARG_ATOM);
+        }
+        int mode = term_to_int(mode_term);
+        fd = open(path, posix_flags, mode);
+    } else {
+        fd = open(path, posix_flags);
+    }
+    free(path);
+    if (UNLIKELY(fd < 0)) {
+        // Return an error.
+        if (UNLIKELY(memory_ensure_free_opt(ctx, TUPLE_SIZE(2), MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
+            RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+        }
+        result = term_alloc_tuple(2, &ctx->heap);
+        term_put_tuple_element(result, 0, ERROR_ATOM);
+        term_put_tuple_element(result, 1, posix_errno_to_atom(errno, glb));
+    } else {
+        // Return a resource object
+        struct PosixFd *fd_obj = enif_alloc_resource(glb->posix_fd_resource_type, sizeof(struct PosixFd));
+        if (IS_NULL_PTR(fd_obj)) {
+            RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+        }
+        fd_obj->fd = fd;
+        if (UNLIKELY(memory_ensure_free_opt(ctx, TUPLE_SIZE(2) + TERM_BOXED_RESOURCE_SIZE, MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
+            RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+        }
+        term obj = term_from_resource(fd_obj, &ctx->heap);
+        enif_release_resource(fd_obj);
```

Comment:
I'm not sure I am properly understanding this code: in my mind here `fd_obj` decreses to 0 refs.

Reply:
refc was incremented in term_from_resource. It's more obvious with enif_* functions as this is their semantics. I'm moving the increment to enif_make_resource wrapper and removing it here.